partner_communication/models/communication_job.py

```diff
@@ -378,6 +370,24 @@ def write(self, vals):
         if vals.get("body_html") or vals.get("send_mode") == "physical":
             self.count_pdf_page()
 
+        # if the call must be done after the sending, we unlink all activities
+        # associated with the job (as for the moment, there is only one activity type)
+        if self.need_call == "after_sending" and self.state == "pending":
+            self.activity_ids.unlink()
+
+        # if the call must be done before the sending, and if there isn't an activity
+        # already scheduled, we schedule a new activity
+        # (there's only one activity type associated with the communication job)
+        if ((self.need_call == "before_sending"
+             and self.state == "pending")
+                and not self.activity_ids):
+            self.activity_schedule('mail.mail_activity_data_call',
+                                   summary="call_" + self.need_call,
+                                   user_id=self.user_id.id,
+                                   note="Call {} at (phone) {}"
+                                   .format(self.partner_id.name,
+                                           self.partner_phone))
```

Comment:
Maybe you should put all this code inside a if condition to execute it only if the `need_call` or `state` fields values are changed. Otherwise it will execute this each time we write any field values in the communication.
```python
        if "need_call" in vals or "state" in vals:
```

Reply:
Is this the write method you mentioned when you talked about the lines 422 to 429 which are in the send method ? Still, I followed your advice and check if the `need_call` and `state` fields values changed and then did a for loop inside `self`.